pkg/controller/servicecidrs/servicecidrs_controller.go

```diff
@@ -0,0 +1,566 @@
+	cidrList, err := c.serviceCIDRLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	// track the names of the different ServiceCIDRs, there
+	// can be multiple ServiceCIDRs sharing the same prefixes
+	// and this is important to determine if a ServiceCIDR can
+	// be deleted.
+	tree := iptree.New[sets.Set[string]]()
+	for _, cidr := range cidrList {
+		if prefix, err := netip.ParsePrefix(cidr.Spec.IPv4); err == nil { // if is empty err will not be nil
+			// if the prefix already exist append the new ServiceCIDR name
+			v, ok := tree.GetPrefix(prefix)
+			if !ok {
+				v = sets.Set[string]{}
+			}
+			v.Insert(cidr.Name)
+			tree.InsertPrefix(prefix, v)
+		}
+		if prefix, err := netip.ParsePrefix(cidr.Spec.IPv6); err == nil { // if is empty err will not be nil
+			// if the prefix already exist append the new ServiceCIDR name
+			v, ok := tree.GetPrefix(prefix)
+			if !ok {
+				v = sets.Set[string]{}
+			}
+			v.Insert(cidr.Name)
+			tree.InsertPrefix(prefix, v)
+		}
+	}
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+	c.tree = tree
+	return nil
+}
+
+func (c *Controller) sync(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+	startTime := time.Now()
+	defer func() {
+		logger.V(4).Info("Finished syncing ServiceCIDR)", "ServiceCIDR", key, "elapsed", time.Since(startTime))
+	}()
+
+	// TODO(aojea) verify if this present a performance problem
+	// restore the radix tree from the current state
+	err := c.syncCIDRs()
+	if err != nil {
+		return err
+	}
+
+	logger.V(4).Info("syncing ServiceCIDR", "ServiceCIDR", key)
+	cidr, err := c.serviceCIDRLister.Get(key)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(4).Info("ServiceCIDR no longer exist", "ServiceCIDR", key)
+			return nil
+		}
+		return err
+	}
+
+	// Deleting ....
+	if !cidr.GetDeletionTimestamp().IsZero() {
+		// check if the existing ServiceCIDR can be deleted before removing the finalizer
+		ok, err := c.canDeleteCIDR(ctx, cidr)
+		if err != nil {
+			return err
+		}
+		// if there are no IPAddress depending on this ServiceCIDR is safe to remove it
+		if ok {
+			return c.removeServiceCIDRFinalizerIfNeeded(ctx, cidr)
+		}
+		// update the status to indicate why the ServiceCIDR can not be deleted
+		svcApplyStatus := networkingapiv1alpha1apply.ServiceCIDRStatus().WithConditions(
+			metav1apply.Condition().
+				WithType(networkingapiv1alpha1.ServiceCIDRConditionReady).
+				WithStatus(metav1.ConditionFalse).
+				WithReason(networkingapiv1alpha1.ServiceCIDRReasonTerminating).
+				WithMessage("There are still IPAddresses referencing the ServiceCIDR, please remove them or create a new ServiceCIDR").
+				WithLastTransitionTime(metav1.Now()))
+		svcApply := networkingapiv1alpha1apply.ServiceCIDR(cidr.Name).WithStatus(svcApplyStatus)
+		_, err = c.client.NetworkingV1alpha1().ServiceCIDRs().ApplyStatus(ctx, svcApply, metav1.ApplyOptions{FieldManager: controllerName, Force: true})
+		return err
```

Comment:
Can you comment here about what will trigger this to be re-evaluated?  Is it a timer or just teh regular periodic full sync?  or something else?

Reply:
it is a trigger , the handlers process  ServiceCIDR or IPAddress changes and enqueue the impacted ServiceCIDRs, just to check that change impact on the state of a ServiceCIDR that is going to be deleted